### DISCUSSION TITLE
Overriding handler / `interpose`ing: how to fordward actions?

### DISCUSSION THREAD

**Discussion by MuhammedZakir** [haskell-effectful/effectful#87] [Q&A] [answered]
For this effect:

```haskell
data E :: Effect where
  Op  :: E m ()
  Op2 :: Eff (E ': es) a -> E m ()
```

when `Op2` is called, I want to handle the `E` effect inside it using a different handler, something like
```haskell
runE = interpret $ \_ -> \case
  ...
  Op2 action -> interpose $ \env act -> liftIO (print "augmented) >> send act
  ...
```

To be precise, in this case, I want to send a signal when any `E` effect is handled inside `action`, without affecting existing handlers.

If above is not possible, I am fine with this (if it will work):
```haskell
runE =  <real handling here>

runIfInsideOp2 = interpose $ \_ act ->
  if insideOp2
  then liftIO (putStrLn "signal") >> send act
  else send act

finalIO = runEff . <other handlers> . runE . runIfInsideOp2
```

However, for above, I am getting this error:
```
• Couldn't match type ‘localEs’ with ‘es’
      Expected: e0 (Eff es) a1
        Actual: e0 (Eff localEs) a1
  --snip--
• In the first argument of ‘send’, namely 'x'
  --snip--
```

---

Edit: Thanks a lot for this library. It is great! :-D

**Comment by arybczak** [comment, marked as answer]
> Thanks a lot for this library. It is great! :-D

Thanks :slightly_smiling_face: 

I'm not sure what you're trying to do and the code doesn't really tell me this (the snippets are not valid code that can generate relevant compile errors). Can you elaborate and provide better code samples?

**Reply by arybczak** [reply]
Perhaps something like this will work for you:

```haskell
data E :: Effect where
  Op  :: E m ()
  Op2 :: Eff (E : es) a -> E (Eff es) a
type instance DispatchOf E = Dynamic

runE :: Eff (E : es) a -> Eff es a
runE = interpret $ \env -> \case
  Op    -> pure ()
  Op2 m -> localSeqUnlift env $ \unlift -> do
    unlift $ runInnerE m

runInnerE :: Eff (E : es) a -> Eff es a
runInnerE = error "runInnerE"
```

**Reply by arybczak** [reply]
With #88 you can do this:

```haskell
data E :: Effect where
  Op  :: E m ()
  Op2 :: m a -> E m a
type instance DispatchOf E = Dynamic

runE :: forall a es. IOE :> es => Eff (E : es) a -> Eff es a
runE = interpret $ \env -> \case
  Op    -> pure ()
  Op2 m -> localLiftUnlift env SeqUnlift $ \lift unlift -> do
    unlift $ interpose (augment lift) m
  where
    augment
      :: E :> localEs
      => (forall r. Eff es r -> Eff localEs r)
      -> EffectHandler E localEs
    augment lift env = \case
      Op    -> do
        lift . liftIO $ putStrLn "Op"
        send Op
      Op2 m -> do
        lift . liftIO $ putStrLn "Op2"
        localSeqUnlift env $ \unlift -> do
          send . Op2 $ unlift m
```

but the problem (apart from this being quite convoluted) might be nested calls to `Op2`, each will add another layer of signaling.

It looks to me like instead of having `Op2` as handler operation you might want to write another handler that `interpose`s and use it explicitly, like this:

```haskell
data E :: Effect where
  Op  :: E m ()
type instance DispatchOf E = Dynamic

runE :: Eff (E : es) a -> Eff es a
runE = interpret $ \env -> \case
  Op -> pure ()

-- Takes over the function of `Op2`.
augmentedE :: (IOE :> es, E :> es) => Eff es a -> Eff es a
augmentedE = interpose $ \env -> \case
  Op -> do
    liftIO $ putStrLn "Op"
    send Op
```

**Reply by MuhammedZakir** [reply]
Edit: Sorry didn't see your posts. I will check them now.

---

The code I posted above is what I am hoping to do, although it is invalid. Sorry for not being clear.

---

I have posted a sample code that is hopefully more clear. Important thing is the interpretation for `LocalS`:

1. for the "action" given using the `localS action` call,
2. I want to execute that "action" using a different handler for `E` effect.
3. That handler does only one thing: if the "action" calls any of the operations of `E` effect, the new handler will first call a function (in the below code, print "augmented"), and then **forwards** those calls to the real/existing handler. In other words, the "action" should work just like it would if the new handler didn't exist.

Why I want to do this: The "action" given will be run in a different thread, and I want to cancel it if doesn't call any of the operations in `E` effect in X milliseconds. I want to create a new handler for effect `E`, so using it, I can send a signal to the original thread when any operation is called. I know I can do it differently, without creating a new handler , but this timeout-like behavior only pertains to 1 or 2 operations (in the below code, `LocalS`), so I feel like using a new handler to send a signal is better for this.

```haskell
module E where

import Effectful
import Effectful.Dispatch.Dynamic
import Effectful.State.Static.Local
import Effectful.TH (makeEffect)


data E :: Effect where
  GetS   :: E m String
  PutS   :: String          -> E m ()
  PrintS :: String          -> E m ()
  LocalS :: Eff (E ': es) a -> E m ()

makeEffect ''E

runE :: (IOE :> es) => [String] -> Eff (E ': es) a -> Eff es [String]
runE strings = reinterpret (execState strings) \_ -> \case
  GetS -> do
    str <- gets head
    modify tail
    pure str

  PutS str -> modify (str :)

  PrintS str -> liftIO . putStrLn $ "send: " <> str

  LocalS action -> (interpose $ \_ e ->
    liftIO (putStrLn "augmented") >> send e) action
    -- what should I actually do here?


run = runEff . runE (fmap show [1..5]) $ action

action = do
  s <- getS

  localS $ do
    putS $ s <> s
    s' <- getS
    printS $ "inside: " <> s'

  s' <- getS
  printS $ "outside: " <> s'
```

**Reply by MuhammedZakir** [reply]
I found one thing that's causing me problem:

In your examples, for the new handler, you're writing the handling of each operation manually (which looks like it should work fine), but I want to do it in a general way:
```
LocalS action -> (interpose $ \_ e ->
  liftIO (putStrLn "augmented") >> send e) action
```

As you can see, I am not doing anything with "e", just forwarding, whereas, in your code, you're "handling" it.

**Reply by arybczak** [reply]
Btw,

> `LocalS :: Eff (E ': es) a -> E m ()`

This doesn't do what you think it does. `es` is existential so you won't be able to do anything with the argument. It should be `LocalS :: Eff (E ': es) a -> E (Eff es) ()` or `LocalS :: m a -> E m ()` (depends if you want to replace the existing handler with interpose or introduce another one).

> In your examples, for the new handler, you're writing the handling of each operation manually (which looks like it should work fine), but I want to do it in a general way:

This can be done with the help of `coerce`, but will only work (i.e. compile) for first order effects, you can't do that for higher order effects, they need to be explicitly unlifted.

```haskell
import Data.Coerce

augmentedE :: (IOE :> es, E :> es) => Eff es a -> Eff es a
augmentedE = interpose @E $ \env op -> do
  liftIO $ putStrLn "augmented"
  send @E $ coerce op
```

**Reply by MuhammedZakir** [reply]
Is the change in PR needed for `coercing`?

When `coerce`ing [1], this signature is the only one that's working: `Eff (E ': es) a  -> E m ()`. Both of the signatures you gave above (for `LocalS`) don't work. However, I won't be able to interpret the action I sent in the original handler. For that, changes in your PR is needed?

[1]
```haskell
runIE :: (IOE :> es, E :> es) => Eff es a -> Eff es a
runIE = interpose @E $ \env op -> do
  liftIO $ putStrLn "augmented"
  send @E $ coerce op
```

Edit: The error I am getting when using those 2 signatures is
```
1. • Couldn't match type ‘localEs’ with ‘es’ arising from a use of ‘coerce’
     ‘localEs’ is a rigid type variable bound by a type expected by the context:
         EffectHandler E es
       at /test.hs:(27,24)-(29,21)
     ‘es’ is a rigid type variable bound by the type signature for:
         runIE :: forall (es :: [Effect]) a. (IOE :> es, E :> es) => Eff es a -> Eff es a
       at /test.hs:26:1-53
```

**Reply by arybczak** [reply]
> Is the change in PR needed for coercing?

No. It will not work with your `E` because it's higher order, you need to match on every operation just as I did [above](https://github.com/haskell-effectful/effectful/discussions/87#discussioncomment-3413673).

**Reply by MuhammedZakir** [reply]
That's a bummer. I guess what I want is not possible atm. Is it that Haskell doesn't support this, or does it have to do with the library (not implemented, a side-effect of the design, etc.)?

Regardless, thanks for your time! :-)

**Reply by arybczak** [reply]
This seems to work (it needs #88 to compile):

```haskell
data E :: Effect where
  GetS   :: E m String
  PutS   :: String -> E m ()
  PrintS :: String -> E m ()
  LocalS :: m a -> E m a
type instance DispatchOf E = Dynamic

runE :: forall a es. (IOE :> es) => [String] -> Eff (E : es) a -> Eff es [String]
runE strings = reinterpret (execState strings) $ \env -> \case
  GetS -> do
    str <- gets head
    modify @[String] tail
    pure str

  PutS str -> modify (str :)

  PrintS str -> liftIO . putStrLn $ "send: " <> str

  LocalS action -> localLiftUnlift env SeqUnlift $ \lift unlift -> do
    unlift $ interpose (augment lift) action

  where
    augmentation :: Eff (State [String] : es) ()
    augmentation = do
      liftIO $ putStrLn "augmented"

    augment
      :: E :> localEs
      => (forall r. Eff (State [String] : es) r -> Eff localEs r)
      -> EffectHandler E localEs
    augment lift env = \case
      GetS          -> lift augmentation >> send GetS
      PutS str      -> lift augmentation >> send (PutS str)
      PrintS str    -> lift augmentation >> send (PrintS str)
      LocalS action -> do
        lift augmentation
        localSeqUnlift env $ \unlift -> do
          send . LocalS $ unlift action

run :: IO [String]
run = runEff . runE (fmap show [1..5::Int]) $ action
  where
    action = do
      s <- send GetS

      send . LocalS $ do
        send . PutS $ s <> s
        s' <- send GetS
        send . PrintS $ "inside: " <> s'

      s' <- send GetS
      send . PrintS $ "outside: " <> s'
```

Isn't this what you want? The `augment` function needs to case-split to handle the inner `LocalS` properly, but that's it.

Note that with this and nested calls to `LocalS` you will get multiple augmentations, if you just want one level of augmentation:

```haskell
    augment lift env = \case
      GetS          -> lift augmentation >> send GetS
      PutS str      -> lift augmentation >> send (PutS str)
      PrintS str    -> lift augmentation >> send (PrintS str)
      LocalS action -> do
        lift augmentation
        localSeqUnlift env $ \unlift -> unlift action
```

this will do the trick.

**Reply by MuhammedZakir** [reply]
Yeah, that works! ! When I mentioned "what I want is not possible", I was referring to not having to "handle" each operation of the effect manually.

P.S. Is there anything blocking that PR? Also, will you push a new version to hackage after merging, or should I use the repo directly as dep's source?

**Reply by arybczak** [reply]
> Yeah, that works!

Great :+1: 

> When I mentioned "what I want is not possible", I was referring to not having to "handle" each operation of the effect manually.

Yeah, unfortunately you need to handle the operations one by one, because higher order actions need their arguments transformed with the unlifting function and you can't really do that generically.

> Is there anything blocking that PR?

No, I just merged it.

> Also, will you push a new version to hackage after merging, or should I use the repo directly as dep's source?

I'll release a new version in a few days, so you can use the repo directly until then.

**Reply by MuhammedZakir** [reply]
Thank you! :-D
